File: packages/py-parity-checking/src/run-tspice/domains/time.ts
import type {
  StepOutput,
  StepTimeDeltet,
  StepTimeEt2Utc,
  StepTimeScdecd,
  StepTimeScencd,
  StepTimeSce2c,
  StepTimeSce2s,
  StepTimeScs2e,
  StepTimeSct2e,
  StepTimeStr2Et,
  StepTimeTimout,
  StepTimeTkvrsn,
  StepTimeTparse,
  StepTimeTpictr,
  StepTimeTimdefGet,
  StepTimeTimdefSet,
  StepTimeUnitim,
} from "../../case-types.js";
import type { RunTspiceContext } from "../context.js";

type TimeStep =
  | StepTimeStr2Et
  | StepTimeEt2Utc
  | StepTimeTkvrsn
  | StepTimeTimout
  | StepTimeDeltet
  | StepTimeUnitim
  | StepTimeTparse
  | StepTimeTpictr
  | StepTimeTimdefGet
  | StepTimeTimdefSet
  | StepTimeScs2e
  | StepTimeSce2s
  | StepTimeScencd
  | StepTimeScdecd
  | StepTimeSct2e
  | StepTimeSce2c;

/** Execute one `time.*` workflow step in tspice. */
export function runTimeStep(context: RunTspiceContext, step: TimeStep): StepOutput {
  switch (step.op) {
    case "time.str2et":
      return { op: step.op, value: context.spice.raw.str2et(step.time) };

    case "time.et2utc":
      return {
        op: step.op,
        value: context.spice.raw.et2utc(step.et, step.format, step.prec),
      };

    case "time.tkvrsn":
      return {
        op: step.op,
        value: context.spice.raw.tkvrsn(step.item),
      };

    case "time.timout":
      return {
        op: step.op,
        value: context.spice.raw.timout(step.et, step.picture),
      };

    case "time.deltet":
      return {
        op: step.op,
        value: context.spice.raw.deltet(step.epoch, step.eptype),
      };

    case "time.unitim":
      return {
        op: step.op,
        value: context.spice.raw.unitim(step.epoch, step.insys, step.outsys),
      };

    case "time.tparse":
      return {
        op: step.op,
        value: context.spice.raw.tparse(step.timstr),
      };

    case "time.tpictr":
      return {
        op: step.op,
        value: context.spice.raw.tpictr(step.sample, step.pictur),
      };

    case "time.timdef":
      if (step.action === "GET") {
        return { op: step.op, value: context.spice.raw.timdef("GET", step.item) };
      }

      context.spice.raw.timdef("SET", step.item, step.value);
      return { op: step.op, value: null };

    case "time.scs2e":
      return {
        op: step.op,
        value: context.spice.raw.scs2e(step.sc, step.sclkch),
      };

    case "time.sce2s":
      return {
        op: step.op,
        value: context.spice.raw.sce2s(step.sc, step.et),
      };

    case "time.scencd":
      return {
        op: step.op,
        value: context.spice.raw.scencd(step.sc, step.sclkch),
      };

    case "time.scdecd":
      return {
        op: step.op,
        value: context.spice.raw.scdecd(step.sc, step.sclkdp),
      };

    case "time.sct2e":
      return {
        op: step.op,
        value: context.spice.raw.sct2e(step.sc, step.sclkdp),
      };

    case "time.sce2c":
      return {
        op: step.op,
        value: context.spice.raw.sce2c(step.sc, step.et),
      };

    default: {
      const exhaustive: never = step;
      throw new Error(`Unhandled time step: ${JSON.stringify(exhaustive)}`);
    }
  }
}

File: packages/py-parity-checking/sidecars/spiceypy/src/py_parity_sidecar/handlers/time.py
from __future__ import annotations

import spiceypy as sp

from ..models import (
    StepOutput,
    StepTimeDeltet,
    StepTimeEt2Utc,
    StepTimeScdecd,
    StepTimeScencd,
    StepTimeSce2c,
    StepTimeSce2s,
    StepTimeScs2e,
    StepTimeSct2e,
    StepTimeStr2Et,
    StepTimeTimdefGet,
    StepTimeTimdefSet,
    StepTimeTimout,
    StepTimeTkvrsn,
    StepTimeTparse,
    StepTimeTpictr,
    StepTimeUnitim,
    WorkflowStep,
)
from ..runtime import SidecarRuntimeContext

SPICE_OUTPUT_LEN = 4096


def _expect_non_empty(value: str, *, label: str) -> str:
    if value == "":
        raise ValueError(f"{label} must be a non-empty string")
    return value


def run_time_step(step: WorkflowStep, _context: SidecarRuntimeContext) -> StepOutput | None:
    if isinstance(step, StepTimeStr2Et):
        et = float(sp.str2et(step.time))
        return StepOutput(op=step.op, value=et)

    if isinstance(step, StepTimeEt2Utc):
        utc = str(sp.et2utc(step.et, step.format, step.prec))
        return StepOutput(op=step.op, value=utc)

    if isinstance(step, StepTimeTkvrsn):
        if step.item != "TOOLKIT":
            raise ValueError("time.tkvrsn.item must be TOOLKIT")
        value = str(sp.tkvrsn(step.item))
        return StepOutput(op=step.op, value=value)

    if isinstance(step, StepTimeTimout):
        picture = _expect_non_empty(step.picture, label="time.timout.picture")
        value = str(sp.timout(step.et, picture, SPICE_OUTPUT_LEN))
        return StepOutput(op=step.op, value=value)

    if isinstance(step, StepTimeDeltet):
        value = float(sp.deltet(step.epoch, step.eptype))
        return StepOutput(op=step.op, value=value)

    if isinstance(step, StepTimeUnitim):
        value = float(sp.unitim(step.epoch, step.insys, step.outsys))
        return StepOutput(op=step.op, value=value)

    if isinstance(step, StepTimeTparse):
        timstr = _expect_non_empty(step.timstr, label="time.tparse.timstr")
        et, errmsg = sp.tparse(timstr, SPICE_OUTPUT_LEN)
        if str(errmsg).strip() != "":
            raise ValueError(str(errmsg).strip())
        return StepOutput(op=step.op, value=float(et))

    if isinstance(step, StepTimeTpictr):
        sample = _expect_non_empty(step.sample, label="time.tpictr.sample")
        _expect_non_empty(step.pictur, label="time.tpictr.pictur")

        # SpiceyPy does not expose the output-template argument used by tspice's raw
        # API, but CSPICE picture derivation is sample-driven. We keep the same
        # input validation behavior and derive the output picture from sample.
        picture, ok, errmsg = sp.tpictr(sample, SPICE_OUTPUT_LEN, SPICE_OUTPUT_LEN)
        if int(ok) != 1:
            message = str(errmsg).strip()
            raise ValueError(message if message != "" else "tpictr failed")

        return StepOutput(op=step.op, value=str(picture))

    if isinstance(step, StepTimeTimdefGet):
        value = sp.timdef("GET", step.item, SPICE_OUTPUT_LEN)
        return StepOutput(op=step.op, value=str(value))

    if isinstance(step, StepTimeTimdefSet):
        sp.timdef("SET", step.item, SPICE_OUTPUT_LEN, step.value)
        return StepOutput(op=step.op, value=None)

    if isinstance(step, StepTimeScs2e):
        et = float(sp.scs2e(step.sc, step.sclkch))
        return StepOutput(op=step.op, value=et)

    if isinstance(step, StepTimeSce2s):
        sclkch = str(sp.sce2s(step.sc, step.et, SPICE_OUTPUT_LEN))
        return StepOutput(op=step.op, value=sclkch)

    if isinstance(step, StepTimeScencd):
        sclkdp = float(sp.scencd(step.sc, step.sclkch))
        return StepOutput(op=step.op, value=sclkdp)

    if isinstance(step, StepTimeScdecd):
        sclkch = str(sp.scdecd(step.sc, step.sclkdp, SPICE_OUTPUT_LEN))
        return StepOutput(op=step.op, value=sclkch)

    if isinstance(step, StepTimeSct2e):
        et = float(sp.sct2e(step.sc, step.sclkdp))
        return StepOutput(op=step.op, value=et)

    if isinstance(step, StepTimeSce2c):
        sclkdp = float(sp.sce2c(step.sc, step.et))
        return StepOutput(op=step.op, value=sclkdp)

    return None
